### PR TITLE
Policy engine integration with new cp

### PR DESCRIPTION
**Pull request by zivnevo**
Includes:
* API calls to CRUD access policies
* Calling policy engine to authorize outgoing/incoming connections
* Policy engine sets remote peer for outgoing connections
* Calling policy engine to authorize adding bindings and exports
* Informing policy engine of added/removed peers and removed bindings and exports
* Persistent storage of access policies

Does not include:
* Handling load-balancing policies

### CODE REVIEW

File: pkg/controlplane/store/types.go [hunk 1]
```diff
@@ -101,3 +103,23 @@ func NewBinding(binding *api.Binding) *Binding {
...
+	api.Policy
+	// Version of the struct when object was created.
+	Version uint32
+}
+
+// Keys return the keys identifying the policy.
+func (ap *AccessPolicy) Keys() []string {
+	return []string{ap.Name}
+}
+
```
Comment (orozery) [inline pkg/controlplane/store/types.go:118]:
```suggestion
```

This is actually a leftover from the time where a generic store existed.

File: pkg/policyengine/PolicyDispatcher.go [hunk 2]
```diff
@@ -36,8 +39,27 @@ const (
...
+type PolicyDecider interface {
+	AddLBPolicy(lbPolicy *LBPolicy) error
+	DeleteLBPolicy(lbPolicy *LBPolicy) error
+
+	AddAccessPolicy(policy *api.Policy) error
+	DeleteAccessPolicy(policy *api.Policy) error
+
+	AuthorizeAndRouteConnection(connReq *event.ConnectionRequestAttr) (event.ConnectionRequestResp, error)
+
+	AddPeer(peer *api.Peer) error
```
Comment (orozery) [inline pkg/policyengine/PolicyDispatcher.go:51]: These (and similar) functions cannot really fail.
Do we envision any future extension that will make them fail-able?

File: pkg/policyengine/PolicyDispatcher.go [hunk 3]
```diff
@@ -36,8 +39,27 @@ const (
...
+	DeletePeer(name string) error
+
+	AddBinding(imp *api.Binding) (event.Action, error)
```
Comment (orozery) [inline pkg/policyengine/PolicyDispatcher.go:54]: Currently this always returns "allow". Same for `AddExport`.
Do we plan to support policies disallowing this kind of operations?
@elevran @kfirtoledo

**Review by orozery — CHANGES_REQUESTED** [review]

File: cmd/gwctl/subcommand/policy.go [hunk 4]
```diff
@@ -182,26 +194,16 @@ func (o *policyGetOptions) run() error {
...
-	accessPolicies, err := g.GetAccessPolicies()
+	accessPolicies, err := g.Policies.List()
 	if err != nil {
 		return err
 	}
 
 	fmt.Printf("Access policies\n")
-	for d := range accessPolicies {
-		fmt.Printf("Access policy %d: %v\n", d, accessPolicies[d])
+	for d, policy := range *accessPolicies.(*[]api.Policy) {
```
[the remark this answers is not in the document]
Comment (zivnevo) [reply cmd/gwctl/subcommand/policy.go:205]: The current code works.

The suggested change causes a panic
```
panic: interface conversion: interface {} is *[]api.Policy, not []*api.Policy
```
I copied this code from `cmd/gwctl/subcommand/import.go` and modified it to use `api.Policy`.

File: pkg/controlplane/store/types.go (hunk 1 above)
Comment (zivnevo) [reply pkg/controlplane/store/types.go]: Removed. Should I also remove the other implementations for the `Keys()` method?
```golang
func (exp *Export) Keys() []string
func (imp *Import) Keys() []string
func (b *Binding) Keys() []string
```

Comment (orozery) [reply pkg/controlplane/store/types.go]: I've seen you already fixed bugs you found on the way (thanks!).
You can fix that as well, if you like.

File: pkg/policyengine/PolicyDispatcher.go (hunk 3 above)
Comment (zivnevo) [reply pkg/policyengine/PolicyDispatcher.go:54]: (this is something that was possible in the old policy engine, but is currently not implemented in the new policy engine)

**Review by orozery — APPROVED** [review]

**Pull request merged**